###############################################################################
#
#     Snakefile - Allele specific expression using ASElux
#
#
###############################################################################

#shell.prefix("set -eo pipefail; echo BEGIN at $(date); ")
#shell.suffix("; exitstat=$?; echo END at $(date); echo exit status was $exitstat; exit $exitstat")

configfile: "../config/config.yaml"

include: "rules/fastq_prep.smk"
include: "rules/allele_specific.smk"
include: "rules/annotation.smk"

import numpy
import pandas as pd
import os
import json

# -------  LOAD SAMPLE LIST  ---------

sample_file = config['SAMPLE_LIST']
sample_df = pd.read_table(sample_file, sep="\t+", header=0, engine='python')
sampleID = sample_df.sampleID
fastq = sample_df.fastq
print(sample_df.sampleID)

#FILES = json.load(open(config['SAMPLES_JSON']))
MERGE_FILES = json.load(open(config['MERGE_FILES_JSON']))
#print()
ALL_SAMPLES = sorted(MERGE_FILES.keys())
ALL_MERGED_SAMPLES = sorted(MERGE_FILES.keys()) 
#print(ALL_MERGED_SAMPLES)


# -------------  RULES  --------------

localrules: move_and_rename_fastqs, ase_get_SNP_refs, ase_get_GENE_refs

#ase_map_snps_to_genes, ase_cross_ref

rule all:
    input:
        expand("../results/11ASE_MAPPINGS/{sampleID}.ase.txt", sampleID = ALL_MERGED_SAMPLES),
#        expand("../results/14SNP2GENES_GW/chr{CHR}_genes_no_snps.txt", CHR = range(1,23)),
#        "../results/15GETGENELIST/Tucci_2019_imprinted_genes.txt", 
#        "../results/15GETGENELIST/genomewide_genelist.txt"
#        "../results/13CROSSREF/summary_6_ase_imprinting_genes_summary.txt"
#        "../results/15CROSSREF_TUCCI/summary_6_tucci_ase_imprinting_genes_summary.txt"
        "../results/16ISGENEIMPRINTED_TUCCI/imp_test.txt"
